File: WINGS_AI_Bootcamp/Haskell/FigurateNum.hs
import Data.Bits ( (.&.) )

isSquare :: Int -> Bool
isSquare n = (truncate . sqrt $ fromIntegral n) ^ 2 == n

isCube :: Int -> Bool
isCube n = let c = round (fromIntegral n ** (1/3)) in c^3 == n

isPowerOfTwo :: Int -> Bool
isPowerOfTwo n = n > 0 && (n .&. (n - 1)) == 0

main :: IO ()
main = do
  print (isSquare 4)       -- True
  print (isCube 27)        -- True
  print (isPowerOfTwo 4)   -- True

File: WINGS_AI_Bootcamp/scala_Ymeg/figurate_num.scala
object NumberChecks:

  def isSquare(n: Int): Boolean =
    val root = math.sqrt(n.toDouble).toInt
    root * root == n

  def isCube(n: Int): Boolean =
    val root = math.round(math.cbrt(n.toDouble)).toInt
    root * root * root == n

  def isPowerOf2(n: Int): Boolean =
    n > 0 && (n & (n - 1)) == 0

@main def runChecks(): Unit =
  println(NumberChecks.isSquare(25))   // true
  println(NumberChecks.isCube(343))    // true
  println(NumberChecks.isPowerOf2(9))  // false
  println(NumberChecks.isPowerOf2(8))  // true
